src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,315 @@
+  }
+
+  Predicate CanActAsToDLIR(CanActAs can_act_as) {
+    return Predicate("canActAs", {can_act_as.left_principal().name(),
+        can_act_as.right_principal().name()}, kPositive);
+  }
+
+  DLIRAssertion SpokenCanActAsToDLIR(Principal speaker,
+                                  CanActAs can_act_as) {
+    Predicate main_predicate = CanActAsToDLIR(can_act_as);
+    
+    // "canActAs" facts are passed to principals via other canActAs facts in 
+    // essentially the same way as attributes. This function adds extra
+    // rules to pass these around. If the `canActAs` under translation
+    // is `X canActAs Z`, then the rule we need to generate is:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X canActAsZ`
+    // (Where Y is a fresh variable)
+    Principal prin_y = Principal(FreshVar());
+
+    // This is `speaker says Y PredX`
+    Predicate generated_lhs =
+        PushPrin(std::string("says_"), prin_y, main_predicate);
+
+    Predicate y_can_act_as_x = Predicate(std::string("canActAs"),
+        {prin_y.name(), can_act_as.left_principal().name()},
+        Sign::kPositive);
+
+    Predicate speaker_says_y_can_act_as_x =
+        PushPrin(std::string("says_"), speaker, y_can_act_as_x);
+
+    // This is `speaker says X canActAs Z`
+    Predicate speaker_says_x_can_act_as_z =
+        PushPrin(std::string("says_"), speaker, main_predicate);
+
+    // This is the full generated rule:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X canActAs Z`
+    return DLIRAssertion(DLIRCondAssertion(
+        generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                        std::move(speaker_says_x_can_act_as_z)}));
+
+  }
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway, so a vector is used in the
+  // return type.
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIR(
+      Principal speaker, BaseFact base_fact) {
+    return std::visit(
+        raksha::utils::overloaded{
+            // To make the typechecker happy the vectors need to be explicitly
+            // declared.
+            [](Predicate predicate) {
+              std::vector<DLIRAssertion> pred_list = {};
+              return std::make_pair(predicate, pred_list);
+            },
+            [this, &speaker](Attribute attribute) {
+              std::vector<DLIRAssertion> pred_list = 
+                {SpokenAttributeToDLIR(speaker, attribute)};
+              return std::make_pair(AttributeToDLIR(attribute), pred_list);
+            },
+            [this, &speaker](CanActAs canActAs) {
+              std::vector<DLIRAssertion> pred_list =
+                {SpokenCanActAsToDLIR(speaker, canActAs)};
+              return std::make_pair(CanActAsToDLIR(canActAs), pred_list);
+            }},
```

Comment:
Rather than explicitly listing the types and doing almost the same thing with each type, would a generic lambda be better? I would view using `overloaded` in this way as a bit of a later resort; it would be nicer if the same code could uniformly handle these different variants.

Reply:
Hmmmm.... but it's not doing almost the same thing for each type. It calls a different function for each type. There's almost no code reuse opportunity between the variants AFAICT. (Could be I don't really understand your proposal though).

Reply:
Although each case is slightly different, you can view them as "the same" in an abstract sense. I think each of the branches fits the following pattern:

```C++
std::vector<DLIRAssertion> pred_list = SpokenItemToDLIRPredList(speaker, item);
return std::make_pair(ItemToDLIR(item), pred_list);
```

Now you can imagine `SpokenItemToDLIRPredList` as being overloaded in the following fashion:

```C++
std::vector<DLIRAssertion> SpokenItemToDLIRPredList(Principal, Predicate) { return { } }

std::vector<DLIRAssertion> SpokenItemToDLIRPredList(Principal speaker, Attribute attr) {
  return SpokenAttributeToDLIR(speaker, attr);
}

std::vector<DLIRAssertion> SpokenItemToDLIRPredList(Principal speaker, CanActAs caa) {
  return SpokenAttributeToDLIR(speaker, canActAs);
}
```

Other patterns might be cleaner. For instance, you could add a `SpokenCanActAsToDLIRPredList` to each of the classes and just call a single member function on `item`. The exact pattern to be used is up in the air. But the point is, in an abstract sense, these overloads are all the same function, and you might as well use C++'s overload resolution and generic lambda functionality to use less code to do this.

Reply:
OK did this for the one you are commenting on. For the other ones in which, for reasons I can't fathom, the compiler insists the args need to be const refs, I can't get the overloading to work. So the other `visit`s, are still using the `overloaded` thing.